### ISSUE TITLE
MQTT Adapter "Close behavior" in case of hono internal errors

### ISSUE THREAD

**Issue by Christian-Schmid** [eclipse/hono#2064]
The MQTT adapter closes the MQTT connection to the device in case of any errors during publishing telemetry / events / answering commands. This includes also more internal errors, e.g. timeouts on getting the assertion. 
 
As the initial connection establishment with a client is somehow resources intense (SSL Handshake, resource allocation, initial get, assert of device and tenant). Does it really make sense to close the mqtt connection in such error cases? I saw the discussion in [1262](https://github.com/eclipse/hono/issues/1262) and the following [PR](https://github.com/eclipse/hono/pull/1263) to close the connection. 

I'm aware that the specifications says with "[MQTT-4.8.0-2]", that the server should close the connection in case of transient errors, but in scenarios where you have a very large number of devices connected, a short issue e.g. in the registration service could have some cascade effect resulting in the disconnect of a large number of devices. 

In the issue [1262](https://github.com/eclipse/hono/issues/1262) there was a discussion ongoing regarding if this could be configurable per tenant. I think such option would be very nice. 

If the clients of the tenant are implemented in a way where they rely on the PubAck and not only on the client disconnect we still would fulfill the QoS guarantees towards the client:

- **Telemetry with QoS 0**: It could leave the connection open and just drop the telemetry message, which would be in line with the guarantees towards the client.

- **Telemetry with QoS 1**: It could leave the connection open and don't pub ack the message. Client would / could try to resend the message.

- **Event**: Same as Telemetry QoS 1 

For cases in which the cause of the failure is a well-known final cause and the cause is successfully received by the adapter (e.g. device is disabled in registry), the adapter should still close the connection. But for other cases such as timeouts, status codes 500) I would like to have the option to not close the connection.

Thanks
Chris

**Comment by kaniyan** [comment]
@sophokles73 Could you please assign this to me, if no one is already working on it.

**Comment by sophokles73** [comment]
I agree with the handling of unprocessable Telemetry QoS 0 messages being in line with the spec and the guarantees made by Hono.

> Telemetry with QoS 1: It could leave the connection open and don't pub ack the message. Client would / could try to resend the message.

Most people I have talked to believe that this behavior is actually *not* compliant with the spec, i.e. the server only has the options to send a PUBACK or close the connection. We have decided for closing the connection as most MQTT client implementations are not prepared to *not* get a PUBACK in response to a PUBLISH. However, they can always handle the server close the connection.

For events, the situation is even worse FMPOV as a client usually publishes an event for something serious and for which it wants to be sure that the event has been received (and persisted) by Hono. Otherwise, the same arguments against simply not sending a PUBACK apply here as well.

What I can image, though, is providing a separate topic for signalling errors to the device. The device could, for example, subscribe to `tenant/device/errors` and receive error messages indicating the reason for not being able to process a message. The MQTT adapter could then
1. send a PUBACK in response to an unprocessable message,
2. drop the unprocessable message and
3. PUBLISH a corresponding error message to the device.

We could then configure at the tenant level, if the MQTT adapter should close the connection on errors or act as described above.

@calohmn @ctron @dejanb WDYT?

**Comment by calohmn** [comment]
Sounds good. As the error topic, I would suggest using `errors/${tenant-id}/${device-id}` (keeping it in line with the `[event|telemetry|command]/${tenant-id}/${device-id}` topic structure).

**Comment by alinaserg** [comment]
Could you please provide some ideas how these error messages will look like? To make sure that the clients can react on them appropriately.

**Comment by e-grigorov** [comment]
> if the MQTT adapter should close the connection on errors

If the device connection is closed because of Hono internal error and this error is still there. What about the next device attempt to connect? I am asking this question because the device reconnection exponential backoff algorithm usually relies on the connection failure. If the next connection is successful and almost immediately disconnected, it'll end with an infinite loop. Actually, I am curios, did you check it with Eclipse Paho?

> act as described above.

@sophokles73, what is the scope of these errors? Do they represent the Hono state or something like a message status?

**Comment by sophokles73** [comment]
@e-grigorov good point. I guess we will need to differentiate between *transient* errors and more *persistent* errors.

An example for a *transient* error would be a lost connection to the device registry or the AMQP messaging network. Such a connection will be re-established very quickly. Thus, an MQTT client will most likely succeed in performing the same operation again after it has reconnected.

A typical example for a *permanent* error would probably be a gateway trying to send messages on behalf of a non-existing device. In such a case, the MQTT client's attempt to again publish such a message after a reconnect will most likely fail again and again (if the missing device hasn't been registered in the meantime). This is exactly the case that we want to remedy with introducing the option to receive error messages. If the device supports receiving such error messages, then the MQTT adapter would keep up the MQTT connection, discard the (unprocessable) message and send an error message to the gateway indicating the problem so that it can stop sending such messages.

Note that if a device support receiving error messages, we could also keep the MQTT connection in the former (transient) example case and send something like a *503 - try again in 2 secs* message to the device.

**Comment by calohmn** [comment]
Here some ideas towards implementing this feature:

**When to use?**
One option would be to enable the changed error handling (no disconnect, send PUBACK if qos>0) and enable publishing the error message when the same gateway/device, that published the telemetry/event message, has an open subscription on the error topic, possibly on another MQTT connection.
That would mean however, that if the connection with the error topic subscription is interrupted, the error handling would change again and the device would get the (unwanted) disconnects on errors.
To prevent that, and to allow the changed error handling also for devices that don't use the error topic, an explicit flag could be used to force the changed error handling if there is no error topic subscription for the device. I could imagine using a property bag parameter on the telemetry/event topic for that. Something like `ignoreErrors=true`.
Another option would be to use an added, MQTT-specific config option in the tenant configuration. But letting the device define the behaviour via the property bag, instead of having a tenant-wide configuration, sounds like the more flexible option to me, so that I would question whether we need the tenant-wide configuration.

**Error topic**
I would suggest this topic format:
```
error/${tenant-id}/${device-id}/[telemetry|event]/${msg-id}/${error-status}
```
The `msg-id` is the message id (packet id) of the corresponding telemetry/event/command-response message for which the error occurred. In case of a telemetry/command-response Qos 0 message, this ID would be empty.
 In order to still be able to correlate an error message to the corresponding message, we might think about adding support for a `correlationId` property bag parameter for the telemetry Qos0 message (see also https://github.com/eclipse/hono/issues/1276#issuecomment-674095874, where such a parameter would also be needed). 

For a command response message, I would choose the request-id used in the command response topic as correlation id:
`error/${tenant-id}/${device-id}/command-response/${req-id}/${error-status}`
This id is alse set for Qos0 command response messages.

The `error-status` in the error topic corresponds to the error status code of the error, taken from the range of HTTP error status codes.

Specific topic (filters) used for different kinds of devices:
- authenticated Device
  subscription topic filter: `error/#`
  error published on:        `error///[telemetry|event]/${msg-id}/${error-status}`
- unauthenticated Device
  subscription topic filter: `error/${tenant-id}/${device-id}/#`
  error published on:        `error/${tenant-id}/${device-id}/[telemetry|event]/${msg-id}/${error-status}`
- authenticated Gateway
  subscription topic filter: `error/#` (possibly also `error//${device-id}/#` for an individual edge device)
  error published on:        `error//${device-id}/[telemetry|event]/${msg-id}/${error-status}`

**Error message payload**
I would suggest a JSON payload with such a structure:
````
{
    "code": error-code,
    "message": "error-message",
    "timestamp": timestamp,
    "message-id": "message-id of the message that caused the error"
}
````
For a command response message error, I would add an additional "request-id" field.

**Comment by sophokles73** [comment]
@calohmn Sounds good to me, in general. Some questions/concerns I have:
1. It is not entirely clear to me if you want to send errors for QoS 0 messages as well. FMPOV this is not necessary.
2. It seems to me that you want to use the MQTT level packet-id in the topic for sending errors caused by event/telemetry messages but use the application level request ID in the topic for sending errors caused by command response messages. IMHO we should always use the packet ID in order to allow devices to always use the same method of correlating errors with messages sent. The request ID could be added in the payload of the error message (if at all necessary) instead.

**Comment by calohmn** [comment]
> It is not entirely clear to me if you want to send errors for QoS 0 messages as well. FMPOV this is not necessary.

For errors happening when processing QoS 0 messages, the connection is currently disconnected as well.
So, there should be some option to change that.

I would find it consistent, if the existence of an error topic suscription (or the added "ignoreErrors" property bag param) would cause that change in error handling also for QoS 0 messages. And not using that topic to send error messages in that case would seem strange FMPOV. An HTTP device also gets an error response if a QoS 0 request results in an error.

If a device really isn't interested in any errors, it could still use the "ignoreErrors" property bag param and not open an error topic subscription.

> 2. [...]

Concerning usage of packet-id and command request-id: The main issue here is that for QoS 0 messages, there is no packet-id. 

Since command response messages always have the unique *request id*, I thought of always using that in the error topic (not requiring an extra "correlationId" property bag param). 

But yes, this difference probably makes things more complicated (also documentation-wise). So, I would change that in order to also use packet-id or correlationId for command-response error messages, the same as with telemetry and event messages.
A device sending QoS 0 command response messages could then look in the error JSON and use the request-id there for correlation if it doesn't want to use an extra "correlationId" property bag param.

**Comment by e-grigorov** [comment]
Having this error channel per message, how do you envision the client behavior? Let's say, the client sends message1, message2 and message3. The message order is important because there are some dependencies. message2 is reported with an error. What's next? The device cannot retry message2 because message3 was already sent. FMPOV, so called "transient" errors will be hard to be managed. Some Hono internal protection mechanisms will be required.

I have one general note about QoS0, yes, the message can be lost by incident but this doesn't mean that these messages are not important and they can be just dropped. FMPOV, the error feedback channel should cover them.

**Comment by calohmn** [comment]
> message2 is reported with an error. What's next? The device cannot retry message2 because message3 was already sent.

If order is important and retries should be done if needed, I think the client would have to wait for either the PUBACK for message2 or a corresponding error message on the error topic before sending message3.

> FMPOV, so called "transient" errors will be hard to be managed. Some Hono internal protection mechanisms will be required.

What kind of mechanism do you have in mind here?

**Comment by e-grigorov** [comment]
>  I think the client would have to wait for either the PUBACK for message2 or a corresponding error message on the error topic before sending message3.

Basically, this means that there will be a guarantee that the message error will be returned before the PUBACK. I am not sure how such synchronization can be implemented on top of libraries like Paho where we have different callbacks and ACK messages are internally handled.

> What kind of mechanism do you have in mind here?

Most simple one, caching. I am nto sure, most probably, we have such?

**Comment by calohmn** [comment]
> Basically, this means that there will be a guarantee that the message error will be returned before the PUBACK.

We could first publish the error message and let the device acknowledge that. Only after the protocol adapter has received that error message PUBACK (or after some timeout period has elapsed), it would send the PUBACK for the original telemetry/event message. (This would of course only be done if the device has subscribed for errors with QoS 1.)
Would this work for you?

> Most simple one, caching. I am nto sure, most probably, we have such?

I'm not sure which kind of data in terms of caching you mean here. We do use caching for device registration or tenant data for example. That means that the device registry, holding this kind of data, being temporarily not available will likely not cause such a "transient" error because cached data is getting used.

**Comment by calohmn** [comment]
Thinking about it more, the above idea with sequential (error) message PUBACKs looks too complicated.

I would rather suggest the following:
We implement 3 different error handling modes:
- **disconnect on error**: this is the current behaviour and conforming to the MQTT spec. If there is a error topic subscription, the error will be published on that topic.
- **ignore errors**: Errors will not lead to a disconnect, there will also be a PUBACK getting sent for QoS 1 telemetry/event messages. If there is a error topic subscription, the error will be published on that topic.
- **ignore and skip PUBACK**: Errors will not lead to a disconnect, there will be **no** PUBACK for a QoS 1 message. If there is a error topic subscription, the error will be published on that topic.

The `onError` property bag param (with values `disconnect` / `ignore` / `skip_ack`) on a published telemetry/event message topic defines which mode is to be used.
If there is no such param, the error handling mode is 
- either "ignore errors" if there is an error topic subscription,
- or "disconnect on error" if there is no error topic subscription.

@e-grigorov Concerning your case of wanting to redeliver a failed message: You could use the "ignore and skip PUBACK" mode here, meaning you would use QoS 1 and wait for *either* the PUBACK *or* the error message, before sending the next message. If you get an error message, you redeliver the current message. Would this work for you?

**Comment by e-grigorov** [comment]
>@e-grigorov Concerning your case of wanting to redeliver a failed message: You could use the "ignore and skip PUBACK" mode here, meaning you would use QoS 1 and wait for either the PUBACK or the error message, before sending the next message. If you get an error message, you redeliver the current message. Would this work for you?

I am not sure that such flow can be implemented on top of typical MQTT library like Paho. The main reason is that QoS is internally managed by the library and the user doesn't need to take care of this while the error channel will be delivered to the user. Without an internal access to the library, I don't think that this retry can be implemented. It's a big issue because no one will start to implement MQTT client from scratch.

FMPOV, "disconnect on error" and "ignore errors" should do the job. "ignore and skip PUBACK" seems to be too much.

**Comment by kaniyan** [comment]
1. FMPOV we cannot rely only on the packet identifier to correlate the error reasons with that of messages sent.  In the MQTT spec, packet id is defined as a non-zero 16-bit identifier and it is normally an incrementing number and the highest possible value is 65535. When a MQTT connection is opened, packet id starts with a value of 1 and then gets incremented as the messages are being sent. Normally MQTT clients (_Paho_ and _vertx MQTT client_) reset this identifier to 1 once the max value is reached. Let us consider a valid scenario, where a device usually opens a connection, publishes a message and closes the connection. In this case the packet identifier or message id will be always 1. I don't think with this packet identifier, we can correlate well the error messages.  

   IMHO It would be better to have a unique identifier to correlate. If a device intends to know about the error reason, then we could also leave the message id generation to the device. For example, the device could send a message id as a property bag (eg. _event/?id={user defined id}_) along with the message. And this value could be used as part of the error topic or inside the error message. This will also work for QoS 0 messages. 


2. Let us assume that a device publishes a message with any of the below property bags to ignore errors. And the device has already been deleted/disabled or so the tenant. In this case, irrespective of the _ignore error_ property bag, the connection should be closed. IMHO the below property bag should only be applicable to transient errors. IMHO we could use the PR #2198, where the cases such as devices deleted/disabled and tenant deleted/disabled are addressed.

> * **ignore errors**: Errors will not lead to a disconnect, there will also be a PUBACK getting sent for QoS 1 telemetry/event messages. If there is a error topic subscription, the error will be published on that topic.
> 
> * **ignore and skip PUBACK**: Errors will not lead to a disconnect, there will be **no** PUBACK for a QoS 1 message. If there is a error topic subscription, the error will be published on that topic.

**Comment by calohmn** [comment]
> 1. FMPOV we cannot rely only on the packet identifier [...]
IMHO It would be better to have a unique identifier to correlate. If a device intends to know about the error reason, then we could also leave the message id generation to the device.

@kaniyan I wouldn't want to *require* the device to always set an extra message id property bag parameter for the error topic feature. For QoS 1 messages, the MQTT message id would be just fine for the usual case, that the device has opened one MQTT connection for sending messages, and that correlation of telemetry message and corresponding error message is only relevant for a short period of time after the telemetry message got sent.
For the case that the device has opened multiple MQTT connections and that the error topic subscription is done in a separate connection, I can imagine using the MQTT client identifier as prefix for the id used in the topic of the published error message:
`error///[telemetry|event]/${msg-id}/${error-status}` with `msg-id` being `${clientIdentifier}_${packet id}`.

> Let us consider a valid scenario, where a device usually opens a connection, publishes a message and closes the connection. In this case the packet identifier or message id will be always 1. I don't think with this packet identifier, we can correlate well the error messages.

If the client uses the same MQTT connection for sending the telemetry and receiving the error messages, correlation would be done in the context and lifetime of the connection. If different MQTT connections are used, each with their own unique client identifier, then the added clientIdentifier prefix would offer an extra means for correlating the messages. So, I don't see an issue here in both cases.
If the client application really needs to uniquely correlate telemetry and error message across a long time period and multiple subsequent MQTT connections, the client should use the extra `correlation-id` property bag param, which it would also need to use for QoS 0 messages.

> 2, [..] the connection should be closed. [..] IMHO we could use the PR #2198, where the cases such as devices deleted/disabled and tenant deleted/disabled are addressed.

I agree, in such cases the connection should always be closed.

**Comment by kaniyan** [comment]
> If the client uses the same MQTT connection for sending the telemetry and receiving the error messages, correlation would be done in the context and lifetime of the connection. If different MQTT connections are used, each with their own unique client identifier, then the added clientIdentifier prefix would offer an extra means for correlating the messages. So, I don't see an issue here in both cases.

I agree that using the Client Identifier together with the packet id solves the correlation issue to a good extent (as the reset of the packet identifier to value 1 after reaching max value is still possible). 

> If the client application really needs to uniquely correlate telemetry and error message across a long time period and multiple subsequent MQTT connections, the client should use the extra `correlation-id` property bag param, which it would also need to use for QoS 0 messages.

I understood that the above `correlation-id` property bag param refers to an identifier given by a device to correlate the error messages. If so, then we are talking about the same identifier. In the above comment, I mentioned it as `message-id` and the purpose is the same.

**Comment by calohmn** [comment]
> I understood that the above `correlation-id` property bag param refers to an identifier given by a device to correlate the error messages. If so, then we are talking about the same identifier. In the above comment, I mentioned it as `message-id` and the purpose is the same. 

Yes. I think I would rather name the property bag parameter `correlation-id` here as not to imply that this has to be the MQTT message id. That id could also be used for the device-side request/response pattern then (see https://github.com/eclipse/hono/issues/1276#issuecomment-674095874).

**Comment by calohmn** [comment]
> > Concerning your case of wanting to redeliver a failed message: You could use the "ignore and skip PUBACK" 

> I am not sure that such flow can be implemented on top of typical MQTT library like Paho. 

@e-grigorov I don't see why this couldn't be implemented with most if not all Paho clients. I mean after having published a message you typically get a delivery token. On/with that token you can either call a method to wait for the PUBACK (along with a timeout parameter) or you can periodically check if the PUBACK has been received (i.e. the delivery identified by the token is complete). There may even be a method to get informed by the PUBACK via a callback.
And in that application code checking the delivery status you could add checks for whether an error message with corresponding message id has been received in the meantime. If that is the case, you can stop waiting for the PUBACK and redeliver the message. Or am I missing something here and it's different with the particular client you are using?

**Comment by calohmn** [comment]
One correction regarding the above design:
MQTT devices will have to create the error subscription **on the same MQTT connection** that they send the telemetry/event messages on.

Otherwise the connection with the error subscription might be with another MQTT adapter instance and there would be no way transporting the telemetry/event sending error from the adapter instance where it occurred, to the adapter instance where the error subscription is. At least there is currently no mechanism for that.

**Comment by calohmn** [comment]
@kaniyan I would take over implementing the error subscription handling, if you don't mind. Maybe you could afterwards adopt the changes from #2198 concerning always closing the connection in case of a non-transient error (device/tenant/adapter not registered/enabled).

**Comment by kaniyan** [comment]
@calohmn Go ahead with the implementation. I can take over the scenario of identifying the non-transient errors and handling them.

**Comment by calohmn** [comment]
@kaniyan The logic for determining the kind of error on which always to do a disconnect can be added here now:

https://github.com/eclipse/hono/blob/60972d3eb03d2373a9ca37b84ef35c90b67413f2/adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java#L1131-L1132

**Comment by sophokles73** [comment]
@calohmn @kaniyan can this be closed?

**Comment by calohmn** [comment]
I would keep it open to still adopt the changes from #2198 concerning always closing the connection in case of a non-transient error (device/tenant/adapter not registered/enabled).

**Comment by kaniyan** [comment]
I would also keep it open. I will create a PR for closing the connection in case of non-transient errors.

**Comment by kaniyan** [comment]
> I would also keep it open. I will create a PR for closing the connection in case of non-transient errors.

With PR #2539, now the MQTT  connections are closed in case of terminal errors.

This issue is closed via #2508 and #2539.
